libdnf/comps/environment/environment.cpp

```diff
@@ -0,0 +1,277 @@
+#include <libdnf/comps/environment/environment.hpp>
+#include <libdnf/comps/environment/query.hpp>
+#include <libdnf/comps/environment/sack.hpp>
+#include <libdnf/comps/group/group.hpp>
+#include <libdnf/comps/comps.hpp>
+#include <libdnf/comps/comps_impl.hpp>
+#include <libdnf/utils/xml.hpp>
+
+extern "C" {
+#include <solv/knownid.h>
+#include <solv/pool.h>
+#include <solv/repo.h>
+#include <solv/solvable.h>
+#include <solv/dataiterator.h>
+}
+
+#include <string>
+#include <iostream>
+#include <libxml/tree.h>
+
+
+namespace libdnf::comps {
+
+
+Environment::~Environment() {}
+
+
+Environment::Environment(EnvironmentQuery * query) : query(query->get_weak_ptr()) {}
+
+
+void add_solvable_id(Environment & environment, Id solvable_id) {
+    environment.add_environment_id(EnvironmentId(solvable_id));
+}
+
+
+void add_solvable_ids(Environment & environment, std::vector<Id> solvable_ids) {
+    for (Id solvable_id : solvable_ids) {
+        environment.add_environment_id(EnvironmentId(solvable_id));
+    }
+}
+
+
+Environment & Environment::operator+=(const Environment & rhs) {
+    this->environment_ids.insert(this->environment_ids.begin(), rhs.environment_ids.begin(), rhs.environment_ids.end());
+    return *this;
+}
+
+
+std::string lookup_str(Pool * pool, std::vector<EnvironmentId> environment_ids, Id key) {
+    for (EnvironmentId environment_id: environment_ids) {
+        if (pool_lookup_str(pool, environment_id.id, key)) {
+            return pool_lookup_str(pool, environment_id.id, key);
+        }
+    }
+    return "";
+}
+
+
+std::string Environment::get_environmentid() const {
+    std::string solvable_name(
+        lookup_str(query->sack->comps.p_impl->get_pool(), environment_ids, SOLVABLE_NAME));
```

Comment:
This is a common occurrence across the codebase. You store the `query` here and only use it to get to the `Pool` in a really complicated way. I'm working on a PR to move the `Pool` to `Base` and I think we should only store `Base` (a `BaseWeakPtr` in particular) in all these objects, unless there's a reason to store something else. Saying this as a heads-up but also if you feel there could be an issue that we should discuss.

Reply:
I don't see any issue with this. There is also a question if we want a separate pool for comps, but I don't know, I am just letting you know there was a discussion around this.